homeassistant/components/weatherflow/sensor.py

```diff
@@ -0,0 +1,398 @@
+
+    @callback
+    def async_add_sensor(device: WeatherFlowDevice) -> None:
+        """Add WeatherFlow sensor."""
+        LOGGER.debug("Adding sensors for %s", device)
+        async_add_entities(
+            WeatherFlowSensorEntity(
+                device, description, hass.config.units is METRIC_SYSTEM
+            )
+            for description in SENSORS
+            if hasattr(
+                device,
+                description.key if description.attr is None else description.attr,
+            )
+        )
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_{SENSOR_DOMAIN}",
+            async_add_sensor,
+        )
+    )
+
+
+class WeatherFlowSensorEntity(SensorEntity):
+    """Defines a WeatherFlow sensor entity."""
+
+    entity_description: WeatherFlowSensorEntityDescription
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        device: WeatherFlowDevice,
+        description: WeatherFlowSensorEntityDescription,
+        is_metric: bool = True,
+    ) -> None:
+        """Initialize a WeatherFlow sensor entity."""
+        self.device = device
+        if not is_metric and (
+            (unit := IMPERIAL_UNIT_MAP.get(description.native_unit_of_measurement))  # type: ignore[arg-type]
+            is not None
+        ):
+            description.native_unit_of_measurement = unit
+        self.entity_description = description
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self.device.serial_number)},
+            manufacturer="WeatherFlow",
+            model=self.device.model,
+            name=f"{self.device.model} {self.device.serial_number}",
+            sw_version=self.device.firmware_revision,
+            suggested_area="Backyard",
+        )
+        self._attr_name = (
+            f"{self.device.model} {self.device.serial_number} {description.name}"
+        )
+        self._attr_unique_id = f"{DOMAIN}_{self.device.serial_number}_{description.key}"
+
+    @property
+    def last_reset(self) -> datetime | None:
+        """Return the time when the sensor was last reset, if any."""
+        if self.entity_description.state_class == SensorStateClass.TOTAL:
+            return self.device.last_report
+        return None
+
+    @property
+    def native_value(self) -> datetime | StateType:
+        """Return the state of the sensor."""
+        attr = getattr(
+            self.device,
+            self.entity_description.key
+            if self.entity_description.attr is None
+            else self.entity_description.attr,
+        )
+
+        if attr is None:
+            return attr
+
+        if (
+            self.hass.config.units is not METRIC_SYSTEM
+            and (function := self.entity_description.conversion_fn) is not None
+        ) or (function := self.entity_description.value_fn) is not None:
+            attr = function(attr)
```

Comment:
We should just add this in the native way it's coming in and then the user can change themselves on the entities.

Reply:
What I currently have in the codebase works great for US - but when I try to mess with things I'm getting different things than I'd like... 

<img width="318" alt="image" src="https://github.com/home-assistant/core/assets/6491743/77662206-4add-4c51-8c7e-d344024b72a0">

Reply:
Let's take air temp as example.
Let's say the device sends the information as 70 F and you have configured the sensor to be natively using F as unit.
Then HA will convert it for a metric user by itself to 21 C.
Now you're doing all kind of conversions to do that for the user when it should not be needed (HA will take care + the user can always change after).

Or can the device send different depending on the setting so you assume it's configured same as their HA? Can we then check that instead?

Reply:
So why do we have all the conversions for what HA can handle itself?
You should set the right device_class and let HA handle it.

- distance https://github.com/home-assistant/core/pull/75530/files#diff-7a3de8519cb87226d4c9bc0fb03c85f5c359dc51b730258e840b3614d962d019R138
- precipitation https://github.com/home-assistant/core/pull/75530/files#diff-7a3de8519cb87226d4c9bc0fb03c85f5c359dc51b730258e840b3614d962d019R157
- precipitation_intensity https://github.com/home-assistant/core/pull/75530/files#diff-7a3de8519cb87226d4c9bc0fb03c85f5c359dc51b730258e840b3614d962d019R166
- etc. (only some examples above)

We should probably also in some remove the translation key and just use the device_class as it's name.

Reply:
Ok I think I'm following (finally)... I'll ping you when I've made some changes